pkg/csi/service/wcpguest/controller.go

```diff
@@ -428,30 +443,59 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 	ctx = logger.NewContextWithLogger(ctx)
 	log := logger.GetLogger(ctx)
 	log.Infof("ControllerUnpublishVolume: called with args %+v", *req)
+
 	err := validateGuestClusterControllerUnpublishVolumeRequest(ctx, req)
 	if err != nil {
 		msg := fmt.Sprintf("Validation for UnpublishVolume Request: %+v has failed. Error: %v", *req, err)
 		log.Error(msg)
 		return nil, err
 	}
 
+	// Retrieve Supervisor PVC
+	svPVC, err := c.supervisorClient.CoreV1().PersistentVolumeClaims(c.supervisorNamespace).Get(ctx, req.VolumeId, metav1.GetOptions{})
+	if err != nil {
+		msg := fmt.Sprintf("failed to retrieve supervisor PVC %q in %q namespace. Error: %+v", req.VolumeId, c.supervisorNamespace, err)
+		log.Error(msg)
+		return nil, status.Error(codes.Internal, msg)
+	}
+	var isFileVolume bool
+	for _, accessMode := range svPVC.Spec.AccessModes {
+		if accessMode == corev1.ReadWriteMany || accessMode == corev1.ReadOnlyMany {
+			isFileVolume = true
+		}
+	}
+	if isFileVolume {
+		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.FileVolume) {
+			return controllerUnpublishForFileVolume(ctx, req, c)
+		}
+		// Feature is disabled on the cluster
+		return nil, status.Error(codes.InvalidArgument, "File volume not supported.")
+	}
+	return controllerUnpublishForBlockVolume(ctx, req, c)
+}
+
+// controllerUnpublishForBlockVolume is helper method to handle ControllerPublishVolume for Block volumes
+func controllerUnpublishForBlockVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest, c *controller) (
```

Comment:
Nit: publish and unpublish methods per volume type can be within controller struct.